core/commands/files/files.go

```diff
@@ -434,6 +438,314 @@ Examples:
+			))
+		} else {
+			// regular connected DAGService
+			dagserv = nd.DAG
+		}
+
+		root, err := getNodeFromPath(req.Context(), nd, dagserv, path)
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		isDirectory := false
+		// try to decode unixfs
+		if pn, ok := root.(*dag.ProtoNode); ok {
+			unixfs, err := ft.FromBytes(pn.Data())
+			if err == nil {
+				isDirectory = unixfs.GetType() == ft.TDirectory
+			}
+		}
+
+		summary, err := walkBlockStart(res, nd.Context(), dagserv, root, isDirectory, onlySummary)
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		res.Emit(&treeOutput{
+			Hash:    root.Cid().String(),
+			Type:    summaryType,
+			Summary: summary,
+		})
+
+	},
+	Type: treeOutput{},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeEncoder(func(req cmds.Request, w io.Writer, v interface{}) error {
+			output, ok := v.(*treeOutput)
+			if !ok {
+				return e.TypeErr(output, v)
+			}
+
+			if output.Type == itemType {
+				item := output.Item
+
+				if item.Depth > 1 {
+					fmt.Fprint(w, strings.Repeat("│   ", item.Depth-1))
+				}
+
+				if item.Depth > 0 {
+					fmt.Fprint(w, "└── ")
+				}
+
+				if item.Depth == 0 {
+					fmt.Fprintln(w, output.Hash)
+				} else {
+					fmt.Fprintln(w, item.Name)
+				}
+
+				return nil
+			}
+
+			if output.Type == summaryType {
+				summary := output.Summary
+
+				fmt.Fprintf(w, "%d directories, %d files present\n",
+					summary.NbDir,
+					summary.NbFile,
+				)
+
+				if summary.SizeTotal > 0 {
+					fmt.Fprintf(w, "%s present of %s (%.2f%%)",
+						humanize.Bytes(uint64(summary.SizeLocal)),
+						humanize.Bytes(uint64(summary.SizeTotal)),
+						100.0*float64(summary.SizeLocal)/float64(summary.SizeTotal),
+					)
+				} else {
+					fmt.Fprintf(w, "%s present of unknown", humanize.Bytes(uint64(summary.SizeLocal)))
+				}
+
+				fmt.Fprintln(w)
+				return nil
+			}
+
+			return fmt.Errorf("unknow output type")
+		}),
+	},
+}
+
+type walkState struct {
+	res     cmds.ResponseEmitter
+	ctx     context.Context
+	dagserv dag.DAGService
+	node    node.Node
+
+	onlySummary       bool
+	depth             int
+	name              string
+	parentIsDirectory bool
+}
+
+func walkBlockStart(res cmds.ResponseEmitter, ctx context.Context, dagserv dag.DAGService, nd node.Node, isDirectory bool, onlySummary bool) (treeSummary, error) {
+	return walkBlock(walkState{
+		res,
+		ctx,
+		dagserv,
+		nd,
+		onlySummary,
+		0,
+		nd.Cid().String(),
+		isDirectory,
+	})
+}
+
+// Build a summary of a dag while emiting treeOutput with treeItem to display the graph of files/directories
+func walkBlock(state walkState) (treeSummary, error) {
+
+	// Start with the block data size
+	result := treeSummary{
+		Local:     true,
+		SizeLocal: uint64(len(state.node.RawData())),
+	}
+
+	isDirectory := false
+	// try to decode unixfs
+	if pn, ok := state.node.(*dag.ProtoNode); ok {
+		unixfs, err := ft.FromBytes(pn.Data())
+		if err == nil {
```

Comment:
I feel like we should try to rework the logic here to avoid nesting too deeply.

Reply:
I can merge the test in one line `if unixfs, err := ft.FromBytes(pn.Data()); err == nil {` which make it more clearer, but that doesn't change the nesting depth.
I could also use a `goto` to jump out of the unixfs section when the decoding fail. What do you think ?